File: 섹션08. A형 문제 풀이 (~2022 상반기)/25. 이상한 윷놀이 (코드트리, 2019 하반기 오전 2번).cpp
#include <stdio.h>

#define MAX (12 + 5)

#define WHITE (0)
#define RED (1)
#define BLUE (2)

int T;
int N, K;
int MAP[MAX][MAX];

struct HORSE
{
	int r;
	int c;
	int dir;
	int pos; // position
};

HORSE horse[10 + 3];
int hcnt;

int board[MAX][MAX][10 + 3];
int index[MAX][MAX]; 

// -, 오른쪽, 왼쪽, 위쪽, 아래쪽
int dr[] = { 0, 0, 0, -1, 1 };
int dc[] = { 0, 1, -1, 0, 0 };

void input()
{
	scanf("%d %d", &N, &K);

	for (int r = 0; r <= N + 1; r++)
		for (int c = 0; c <= N + 1; c++)
			MAP[r][c] = BLUE;

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			scanf("%d", &MAP[r][c]);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			index[r][c] = 0;

	hcnt = 1; // 1번 말부터 K번 말까지
	for (int k = 0; k < K; k++)
	{
		int r, c, dir;

		scanf("%d %d %d", &r, &c, &dir);

		horse[hcnt].r = r;
		horse[hcnt].c = c;
		horse[hcnt].dir = dir;

		int& position = index[r][c];
		
		board[r][c][position] = hcnt;
		horse[hcnt].pos = position;

		position++;

		hcnt++;
	}
}

void printBoard() // for debug
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
		{
			int idx = index[r][c];
			if (idx == 0) continue;

			printf("%d, %d : ", r, c);
			for (int i = 0; i < idx; i++)
				printf("%d ", board[r][c][i]);
			putchar('\n');
		}
	}
	putchar('\n');
}

void moveWhite(int number)
{
	HORSE h = horse[number];
	int sr, sc, dir;
	int nr, nc;

	sr = h.r;
	sc = h.c;
	dir = h.dir;

	nr = sr + dr[dir];
	nc = sc + dc[dir];

	int front = h.pos;
	int& back = index[sr][sc];
	int& nextBack = index[nr][nc];

	for (int k = front; k < back; k++)
	{
		int selectedNumber = board[sr][sc][k];

		horse[selectedNumber].r = nr;
		horse[selectedNumber].c = nc;
		horse[selectedNumber].pos = nextBack;

		board[nr][nc][nextBack++] = selectedNumber;
	}

	back = front;
}

void moveRed(int number)
{
	HORSE h = horse[number];
	int sr, sc, dir;
	int nr, nc;

	sr = h.r;
	sc = h.c;
	dir = h.dir;

	nr = sr + dr[dir];
	nc = sc + dc[dir];

	int front = h.pos;
	int& back = index[sr][sc];
	int& nextBack = index[nr][nc];

	for (int k = back - 1; k >= front; k--)
	{
		int selectedNumber = board[sr][sc][k];

		horse[selectedNumber].r = nr;
		horse[selectedNumber].c = nc;
		horse[selectedNumber].pos = nextBack;

		board[nr][nc][nextBack++] = selectedNumber;
	}

	back = front;
}

int simulation()
{
	int changeDir[5] = { 0, 2, 1, 4, 3 };

	for (int i = 1; i <= 1000; i++)
	{
		// printBoard();
		// printf("===================================\n");

		// 1번 말 부터 이동
		for (int h = 1; h < hcnt; h++)
		{
			int sr, sc, dir;
			int nr, nc;
			
			sr = horse[h].r;
			sc = horse[h].c;
			dir = horse[h].dir;

			nr = sr + dr[dir];
			nc = sc + dc[dir];

			if (MAP[nr][nc] == WHITE) moveWhite(h);
			else if (MAP[nr][nc] == RED) moveRed(h);
			else if (MAP[nr][nc] == BLUE)
			{
				dir = changeDir[dir];
				horse[h].dir = dir;

				nr = sr + dr[dir];
				nc = sc + dc[dir];

				if (MAP[nr][nc] == WHITE) moveWhite(h);
				else if (MAP[nr][nc] == RED) moveRed(h);
				// BLUE는 아무 일도 하지 않음.
			}

			if (index[nr][nc] >= 4) return i;
		}
	}

	return -1;
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		printf("%d\n", simulation());
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/25. 이상한 윷놀이 (코드트리, 2019 하반기 오전 2번).py
MAX = 12 + 5
WHITE = 0
RED = 1
BLUE = 2

MAP = [[0] * MAX for _ in range(MAX)]

class HORSE:
    def __init__(self, r: int, c: int, dir: int, pos: int):
        self.r = r
        self.c = c
        self.dir = dir
        self.pos = pos # position

horse = [HORSE(0, 0, 0, 0) for _ in range(10 + 3)]
hcnt = 0

board = [[[0] * (10 + 3) for _ in range(MAX)] for _ in range(MAX)]
index = [[0] * MAX for _ in range(MAX)]

# -, 오른쪽, 왼쪽, 위쪽, 아래쪽
dr = [0, 0, 0, -1, 1]
dc = [0, 1, -1, 0, 0]

def input_data():
    global N, K, MAP, hcnt, index

    N, K = map(int, input().split())
    
    MAP = [[BLUE] * MAX for _ in range(MAX)]

    for r in range(1, N + 1):
        MAP[r][1:N + 1] = map(int, input().split())

    index = [[0] * MAX for _ in range(MAX)]

    hcnt = 1 # 1번 말부터 K번 말까지
    for _ in range(K):
        r, c, dir = map(int, input().split())
        
        horse[hcnt] = HORSE(r, c, dir, index[r][c])                
        board[r][c][index[r][c]] = hcnt
        
        index[r][c] += 1
        hcnt += 1

def print_board(): # for debug
    for r in range(1, N + 1):
        for c in range(1, N + 1):
            idx = index[r][c]
            if idx == 0: continue
            
            print(f"{r}, {c} :", end=' ')
            for i in range(idx):
                print(board[r][c][i], end=' ')
            print("")
    print("")

def move_white(number):
    h = horse[number]
    sr, sc, dir = h.r, h.c, h.dir

    nr = sr + dr[dir]
    nc = sc + dc[dir]
    
    front = h.pos
    for k in range(front, index[sr][sc]):
        selected_number = board[sr][sc][k]
        
        horse[selected_number].r = nr
        horse[selected_number].c = nc
        horse[selected_number].pos = index[nr][nc]
        
        board[nr][nc][index[nr][nc]] = selected_number
        index[nr][nc] += 1

    index[sr][sc] = front

def move_red(number):
    h = horse[number]
    sr, sc, dir = h.r, h.c, h.dir

    nr = sr + dr[dir]
    nc = sc + dc[dir]
    
    front = h.pos
    for k in range(index[sr][sc] - 1, front - 1, -1):
        selected_number = board[sr][sc][k]
        
        horse[selected_number].r = nr
        horse[selected_number].c = nc
        horse[selected_number].pos = index[nr][nc]
        
        board[nr][nc][index[nr][nc]] = selected_number
        index[nr][nc] += 1

    index[sr][sc] = front

def simulation():
    change_dir = [0, 2, 1, 4, 3]
    
    for i in range(1, 1001):
        # 1번 말 부터 이동
        for h in range(1, hcnt):
            sr, sc, dir = horse[h].r, horse[h].c, horse[h].dir
            nr = sr + dr[dir]
            nc = sc + dc[dir]

            if MAP[nr][nc] == WHITE: move_white(h)
            elif MAP[nr][nc] == RED: move_red(h)
            elif MAP[nr][nc] == BLUE:
                dir = change_dir[dir]
                horse[h].dir = dir
                
                nr = sr + dr[dir]
                nc = sc + dc[dir]

                if MAP[nr][nc] == WHITE: move_white(h)
                elif MAP[nr][nc] == RED: move_red(h)
                # BLUE이면 이동하지 않음

            if index[nr][nc] >= 4: return i

    return -1

# T = int(input())
T = 1
for _ in range(T): 
    input_data()
    
    print(simulation())